From: Robert Fairley <rfairley@redhat.com>
Date: Wed, 17 Jun 2020 10:14:19 -0400
Subject: [PATCH] examples/nm-dispatcher.dhcp: use sysconfig

Use the PEERNTP and NTPSERVERARGS environment variables from
/etc/sysconfig/network{-scripts}.

Co-Authored-By: Christian Glombek <cglombek@redhat.com>

diff --git a/examples/chrony.nm-dispatcher.dhcp b/examples/chrony.nm-dispatcher.dhcp
index 6ea4c37..a6ad35a 100644
--- a/examples/chrony.nm-dispatcher.dhcp
+++ b/examples/chrony.nm-dispatcher.dhcp
@@ -6,16 +6,24 @@
 
 chronyc=/usr/bin/chronyc
 default_server_options=iburst
-server_dir=/var/run/chrony-dhcp
+server_dir=/run/chrony-dhcp
 
 dhcp_server_file=$server_dir/$interface.sources
 # DHCP4_NTP_SERVERS is passed from DHCP options by NetworkManager.
 nm_dhcp_servers=$DHCP4_NTP_SERVERS
 
+[ -f /etc/sysconfig/network ] && . /etc/sysconfig/network
+[ -f /etc/sysconfig/network-scripts/ifcfg-"${interface}" ] && \
+    . /etc/sysconfig/network-scripts/ifcfg-"${interface}"
+
 add_servers_from_dhcp() {
     rm -f "$dhcp_server_file"
+
+    # Don't add NTP servers if PEERNTP=no specified; return early.
+    [ "$PEERNTP" = "no" ] && return
+
     for server in $nm_dhcp_servers; do
-        echo "server $server $default_server_options" >> "$dhcp_server_file"
+        echo "server $server ${NTPSERVERARGS:-$default_server_options}" >> "$dhcp_server_file"
     done
     $chronyc reload sources > /dev/null 2>&1 || :
 }
-- 
2.29.2

Use chrony-helper instead of chronyc to avoid changes in default chrony.conf

diff -up chrony-4.1/examples/chrony.nm-dispatcher.dhcp.nm-dispatcher-dhcp chrony-4.1/examples/chrony.nm-dispatcher.dhcp
--- chrony-4.1/examples/chrony.nm-dispatcher.dhcp.nm-dispatcher-dhcp	2021-06-09 11:10:30.997416152 +0200
+++ chrony-4.1/examples/chrony.nm-dispatcher.dhcp	2021-06-09 11:16:23.598381336 +0200
@@ -9,11 +9,12 @@ export LC_ALL=C
 interface=$1
 action=$2
 
-chronyc=/usr/bin/chronyc
+helper=/usr/libexec/chrony-helper
 default_server_options=iburst
-server_dir=/run/chrony-dhcp
+server_dir=/run/chrony-helper
 
-dhcp_server_file=$server_dir/$interface.sources
+dhcp_server_tmpfile=$server_dir/tmp-nm-dhcp.$interface
+dhcp_server_file=$server_dir/nm-dhcp.$interface
 # DHCP4_NTP_SERVERS is passed from DHCP options by NetworkManager.
 nm_dhcp_servers=$DHCP4_NTP_SERVERS
 
@@ -24,24 +24,30 @@ nm_dhcp_servers=$DHCP4_NTP_SERVERS
 add_servers_from_dhcp() {
     rm -f "$dhcp_server_file"
 
+    # Remove servers saved by the dhclient script before it detected NM.
+    rm -f "/var/lib/dhclient/chrony.servers.$interface"
+
     # Don't add NTP servers if PEERNTP=no specified; return early.
     [ "$PEERNTP" = "no" ] && return
 
+    # Create the directory with correct SELinux context.
+    $helper create-helper-directory > /dev/null 2>&1
+
     for server in $nm_dhcp_servers; do
-        echo "server $server ${NTPSERVERARGS:-$default_server_options}" >> "$dhcp_server_file"
+        echo "$server ${NTPSERVERARGS:-$default_server_options}" >> "$dhcp_server_tmpfile"
     done
+    [ -e "$dhcp_server_tmpfile" ] && mv "$dhcp_server_tmpfile" "$dhcp_server_file"
-    $chronyc reload sources > /dev/null 2>&1 || :
+
+    $helper update-daemon > /dev/null 2>&1 || :
 }
 
 clear_servers_from_dhcp() {
     if [ -f "$dhcp_server_file" ]; then
         rm -f "$dhcp_server_file"
-        $chronyc reload sources > /dev/null 2>&1 || :
+        $helper update-daemon > /dev/null 2>&1 || :
     fi
 }
 
-mkdir -p $server_dir
-
 if [ "$action" = "up" ] || [ "$action" = "dhcp4-change" ]; then
     add_servers_from_dhcp
 elif [ "$action" = "down" ]; then
